### PR TITLE
refactor(rome_js_analyze): improve noBannedTypes

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

- Improve the rule documentation
- Allow the type `{}` in contexts where it makes sense (type constraints and type intersections).
- Don't report user types named like a banned type

## Test Plan

New tests included

## Changelog

- [x] The PR requires a changelog line

## Documentation

- [ ] The PR requires documentation
- [ ] I will create a new PR to update the documentation

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -70,6 +70,30 @@ parameter decorators:
   - inside an export default function
   - for React.use* hooks
 
+- Relax [`noBannedTypes`](https://docs.rome.tools/lint/rules/nobannedtypes/) and improve documentation
+
+  The rule no longer reports a user-type that reuses a banned type name.
```
Comment (ematipico) [inline CHANGELOG.md:75]:
```suggestion
  The rule no longer reports a user type that reuses a banned type name.
```

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_banned_types.rs [hunk 2]
```diff
@@ -1,21 +1,65 @@
...
+    ///
+    /// - Enforce consistent names for primitive types
+    ///
+    ///   Primitive types have aliases.
+    ///   For example, `Number` is an alias of `number`.
+    ///   The rule recommends the lowercase primitive type names.
+    ///
+    /// - Disallow the `Function` type
+    ///
+    ///   The `Function` is type is loosely typed and is thus considered dangerous or harmful.
```
Comment (ematipico) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_banned_types.rs:28]:
```suggestion
    ///   The `Function` type is loosely typed and is thus considered dangerous or harmful.
```

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_banned_types.rs [hunk 3]
```diff
@@ -53,9 +97,121 @@ declare_rule! {
...
+                // type NonNull<T> = T & {}
+                // ```
+                if ts_object_type.members().is_empty()
+                    && (ts_object_type.parent::<TsTypeConstraintClause>().is_none()
+                        && ts_object_type
+                            .parent::<TsIntersectionTypeElementList>()
+                            .is_none())
+                {
+                    let range = TextRange::new(
+                        ts_object_type
```
Comment (ematipico) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_banned_types.rs:123]: I **think** `ts_object_type.range()` should contain the range you want. Unless I'm missing something...

**Review by ematipico — APPROVED** [review]

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_banned_types.rs (hunk 3 above)
Comment (Conaclos) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_banned_types.rs]: Good catch! I didn't take the time to specifically look at this code.

**Pull request merged**